alternator/streams.cc

```diff
@@ -479,50 +494,125 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
     // or on expired...
     // TODO: maybe add secondary index to topology table to enable this?
     return _sdks.cdc_get_versioned_streams({ tm.count_normal_token_owners() }).then([this, &db, schema, shard_start, limit, ret = std::move(ret), stream_desc = std::move(stream_desc)](std::map<db_clock::time_point, cdc::streams_version> topologies) mutable {
-        auto i = topologies.begin();
+
+        // filter out cdc generations older than the table or now() - dynamo_streams_max_window (24h)
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - dynamo_streams_max_window);
+
+        auto i = topologies.lower_bound(low_ts);
+        // need first gen _intersecting_ the timestamp.
+        if (i != topologies.begin()) {
+            i = std::prev(i);
+        }
+
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // i is now at the youngest generation we include. make a mark of it.
+        auto first = i;
+
+        // if we're a paged query, skip to the generation where we left of.
+        if (shard_start) {
+            i = topologies.find(shard_start->start_time);
+        }
+
+        // need a prev even if we are skipping stuff
+        if (i != first) {
+            prev = std::prev(i);
+            std::stable_sort(prev->second.streams.begin(), prev->second.streams.end(), [](auto& id1, auto& id2) { 
+                return id1.token() < id2.token(); 
+            });
+        }
+
         for (; limit > 0 && i != e; prev = i, ++i) {
             auto& [ts, sv] = *i;
 
             last = std::nullopt;
 
-            for (auto& id : sv.streams()) {
-                if (shard_start && shard_start->id != id) {
-                    continue;
-                }
-                if (shard_start && shard_start->id == id) {
-                    shard_start = std::nullopt;
-                    continue;
+            auto get_expiration = [&](auto iter) {
+                if (++iter != e) {
+                    return iter->first;
                 }
+                return db_clock::time_point{};
+            };
+
+            auto lo = sv.streams.begin();
+            auto end = sv.streams.end();
+
+            // first sort by index.
+            std::stable_sort(lo, end, [](auto& id1, auto& id2) {
+                return id1.index() < id2.index();
+            });
+
+            if (shard_start) {
+                // find next shard position
+                // first id set is the wraparound range. skip it here.
```

Comment:
Isn't that an implementation detail? Are we sure this will be guaranteed forever? If yes, then probably a unit test should check it somehow.

Reply:
It is using inside knowledge, but otoh it is in the same source code base as said knowledge, so...